### PR TITLE
Handling MODE returned by server

### PR DESCRIPTION
**Pull request by elazkani**
Description
------------
Transforming the MODE returned by the server into USERMODE or CHANNELMODE triggers depending on the type of parameters they are.

`NOTE`: This has been tested on ircu2 and derivatives

```
$ coverage run --branch --source=bottom -m py.test --ignore=tests/integ
======================================================== test session starts ========================================================
platform darwin -- Python 3.6.1, pytest-3.2.2, py-1.4.34, pluggy-0.4.0
rootdir: /Users/Elijah/sandbox/bottom, inifile:
collected 107 items

tests/unit/test_client.py ....................
tests/unit/test_pack.py .................................................
tests/unit/test_protocol.py .........
tests/unit/test_unpack.py .............................

==================================================== 107 passed in 0.52 seconds =====================================================
```

Test code
----------
```python
    @bot.on("USERMODE")
    def umode(**kwargs):
        logger.debug("USERMODE {}".format(kwargs))

    @bot.on("CHANNELMODE")
    def cmode(**kwargs):
        logger.debug("CHANNELMODE {}".format(kwargs))
```

Sample output:
```
USERMODE {'nick': 'nck', 'user': 'usr', 'host': 'hst', 'modes': '+iwx'}
CHANNELMODE {'nick': 'nck', 'user': 'usr', 'host': 'hst', 'channel': '#ch', 'modes': '+o', 'params': 'bot'}
```

### CODE REVIEW

**Comment by numberoverzero** [comment]
Thanks for the PR!  On first review it looks good but I remember some grumbling about MODE before.  Maybe it's easier than I thought.

Historically it's been about a week until I get through a PR; that's not for lack of interest, just a limitation of free time.

**Comment by elazkani** [comment]
@numberoverzero that's not a problem. I tried to keep it consistent with the `pack` function. I just noticed that Travis is failing due to flake8, maybe I did not test it enough. I'll try to fix that issue and apply another commit.

**Comment by elazkani** [comment]
I found out 2 edge cases that I added. Now the output will look something like the following

```
CHANNELMODE {'nick': 'nck', 'user': 'usr', 'host': 'hst', 'channel': '#ch', 'modes': '+m', 'params': ''}
CHANNELMODE {'nick': 'nck', 'user': 'usr', 'host': 'hst', 'channel': '#ch', 'modes': '+o', 'params': ['params']}
```

And I only returned it in this format to keep compatibility with previous code.

File: bottom/unpack.py [hunk 1]
```diff
@@ -300,6 +300,21 @@ def unpack_command(msg: str) -> Tuple[str, Dict[str, Any]]:
         else:
             kwargs["message"] = ""
 
+    elif command in ["MODE"]:
+        nickmask(prefix, kwargs)
+        if "#" in params[0]:
```
Comment (numberoverzero) [inline bottom/unpack.py]: This is simply a convention, right?  Do we have any stronger way to tell apart user/channel mode commands?  For example, minimum arg lengths...

Here are the examples from [rfc 2812](https://www.rfc-editor.org/rfc/rfc2812.txt).

3.1.5 User mode message

```
(1)MODE WiZ -w                     ; Command by WiZ to turn off
                                   reception of WALLOPS messages.

(2)MODE Angel +i                   ; Command from Angel to make herself
                                   invisible.

(3)MODE WiZ -o                     ; WiZ 'deopping' (removing operator
                                   status).
```

3.2.3 Channel mode message

```
(1)MODE #Finnish +imI *!*@*.fi     ; Command to make #Finnish channel
                                   moderated and 'invite-only' with user
                                   with a hostname matching *.fi
                                   automatically invited.

(2)MODE #Finnish +o Kilroy         ; Command to give 'chanop' privileges
                                   to Kilroy on channel #Finnish.

(3)MODE #Finnish +v Wiz            ; Command to allow WiZ to speak on
                                   #Finnish.

(4)MODE #Fins -s                   ; Command to remove 'secret' flag
                                   from channel #Fins.

(5)MODE #42 +k oulu                ; Command to set the channel key to
                                   "oulu".

(6)MODE #42 -k oulu                ; Command to remove the "oulu"
                                   channel key on channel "#42".
```


I think there are completely ambiguous cases eg. 3.1.5 (1) and 3.2.3 (4) are both `MODE [^\s]+ [^\s]+`, in which case we either:
1. Codify the convention that any first parameter starting with the literal "#" is a channel command; user otherwise.
2. Leave it up to the user to parse, and only expose one command `@client.on("MODE")`.

For the first I decided to see if channels or nicks have specific constraints.  There's this bit, from "1.3 Channels":

> 1.3 Channels
> 
>    Channels names are strings (beginning with a '&', '#', '+' or '!'
>    character) of length up to fifty (50) characters.  Apart from the
>    requirement that the first character is either '&', '#', '+' or '!',
>    the only restriction on a channel name is that it SHALL NOT contain
>    any spaces (' '), a control G (^G or ASCII 7), a comma (',').  Space
>    is used as parameter separator and command is used as a list item
>    separator by the protocol).  A colon (':') can also be used as a
>    delimiter for the channel mask.  Channel names are case insensitive.

So I'm ok with the first option (splitting the command into two modes based on `params[0]`) as long as we check that `params[0] in "&#+!"`.  If that's too restrictive, I'm +1 for the second option of simply exposing `MODE` as a single event that unpacks to a dict with keys `[*nickmask, target, modes, params]` where target is either a nickname or channel.

File: bottom/unpack.py [hunk 2]
```diff
@@ -300,6 +300,21 @@ def unpack_command(msg: str) -> Tuple[str, Dict[str, Any]]:
...
+            command = "CHANNELMODE"
+            kwargs["channel"] = params[0]
+            kwargs["modes"] = params[1]
+            if len(params) > 2:
+                kwargs["params"] = params[2:]
+            else:
+                kwargs["params"] = ""
```
Comment (numberoverzero) [inline bottom/unpack.py]: Shouldn't we have an empty list here so `kwargs["params"]` is always a list?  I think (sincerely hope) that's the current convention.

File: tests/unit/test_unpack.py [hunk 3]
```diff
@@ -183,6 +183,41 @@ def test_topic():
...
+    expected_kwargs['nick'] = 'nck'
+    assert set(expected_kwargs) == set(parameters(command))
+
+
+def test_channelmode():
+    """ MODE command """
+    command = "CHANNELMODE"
+    expected_kwargs = {"channel": "#ch", "host": "",
+                       "modes": "+o", "params": ['trget']}
+    message = "MODE #ch +o trget"
```
Comment (numberoverzero) [inline tests/unit/test_unpack.py]: Please make this test expect multiple params, eg:

```MODE #ch +o trget trget2```

should generate

```kwargs["params"] == ["trget", "trget2"]```

File: bottom/unpack.py (hunk 2 above)
Comment (elazkani) [reply bottom/unpack.py]: That's fine, I'll change this to an empty list to keep the convention.

File: tests/unit/test_unpack.py (hunk 3 above)
Comment (elazkani) [reply tests/unit/test_unpack.py]: No problem.

File: bottom/unpack.py (hunk 1 above)
Comment (elazkani) [reply bottom/unpack.py]: I only split the MODE to keep it consistent with the library return. We can keep it the same way, as IRC clients do, and return MODE. Imho I think it's a good idea to keep it consistent in a way that pack and unpack work with the same way. We can send MODE and trigger MODE or do what we did and split them into 2. There are a lot of ambiguous cases in the RFC where it keeps things unclear.

**Comment by elazkani** [comment]
So I bumped the version as well so that people using this will not break their stuff. I wasn't sure if you wanted to bump the minor or the major so I stuck with the minor. Also, forgive the typo in the commit message, I don't want to fix it by rewriting history that would cause travis to go crazy, it can be fixed by squashing before merging I guess. =)

**Comment by numberoverzero** [comment]
Thanks for your contribution!
